### PR TITLE
Change: make as many savegame chunks a CH_ARRAY as possible

### PR DESCRIPTION
**Pull request by TrueBrain**
## Motivation / Problem

Weird goal, I know right? But hear me out.

#9322 sets out to make to make our savegame self-descriptive (see motivation for that there). But in order to do that, we need to make all chunks a table.

The first step in making anything a table, is to first make it an array. And so we come to the motivation of this PR: to make that possible.

## Description

Most `CH_RIFF`s in fact are just an array of length 1. So it might seem silly at first, but that is exactly what this PR does: make those chunks into an array of length 1.

The exception was the `GLOG` chunk. That required some more love and attention.

After this PR, only the map-chunks are `CH_RIFF`. Those could be made into `CH_ARRAY`, but that is not really useful, as they cannot become self-describing for now anyway.

Lastly, the chunk table tells how chunks are stored on disk. Chunks that are no longer stored were still marked by their old type. This was often confusing. So I sneaked that change in here, to mark those as `CH_READONLY`, making it more clear that it is not by accident.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/saveload/animated_tile_sl.cpp [hunk 1]
```diff
@@ -18,13 +18,17 @@
 
 extern std::vector<TileIndex> _animated_tiles;
 
+static const SaveLoad _animated_tile_desc[] = {
```
Comment (rubidium42) [inline src/saveload/animated_tile_sl.cpp:21]: In the commit message: practicle -> practical

File: src/saveload/animated_tile_sl.cpp [hunk 2]
```diff
@@ -45,18 +49,25 @@ static void Load_ANIT()
 		return;
 	}
 
-	uint count = (uint)SlGetFieldLength() / sizeof(_animated_tiles.front());
-	_animated_tiles.clear();
-	_animated_tiles.resize(_animated_tiles.size() + count);
-	SlCopy(_animated_tiles.data(), count, SLE_UINT32);
+	if (IsSavegameVersionBefore(SLV_RIFF_TO_ARRAY)) {
+		uint count = (uint)SlGetFieldLength() / sizeof(_animated_tiles.front());
```
Comment (rubidium42) [inline src/saveload/animated_tile_sl.cpp]: YIKYOMI, but count can be `size_t` as everything using it accepts `size_t`

File: src/saveload/saveload.cpp [hunk 3]
```diff
@@ -1444,6 +1445,48 @@ static void SlDeque(void *deque, VarType conv)
...
+		default: NOT_REACHED();
+	}
+}
+
+/**
+ * Save/load a std::vector.
+ * @param vector The std::vector being manipulated
+ * @param conv VarType type of variable that is used for calculating the size
+ */
+static void SlVector(void *vector, VarType conv)
```
Comment (rubidium42) [inline src/saveload/saveload.cpp:1482]: I don't think it was meant that this got added in this commit.

Comment (TrueBrain) [reply src/saveload/saveload.cpp:1482]: It was, as this is the first use. I can split it off in a new commit that only adds it, but doesn't use it. Amount of code was low enough that I personally didn't care :)

File: src/saveload/gamelog_sl.cpp [hunk 4]
```diff
@@ -15,147 +15,317 @@
...
+		if (lc->ct != GLCT_GRFBUG) return;
+		SlObject(lc, this->GetDescription());
+	}
+
+	void Save(LoggedChange *lc) const override { this->GenericSaveLoad(lc); }
+	void Load(LoggedChange *lc) const override { this->GenericSaveLoad(lc); }
+	void LoadCheck(LoggedChange *lc) const override { this->GenericSaveLoad(lc); }
 };
 
-static_assert(lengthof(_glog_desc) == GLCT_END);
```
Comment (rubidium42) [inline src/saveload/gamelog_sl.cpp:92]: Would it make sense to add something similar to `SlGamelogAction::description` but with a `+ 1`? Alternatively you could say that if someone adds a game log action but do not check it survives a save-load, then the game log action is pointless and should not be added.

**Review by rubidium42 — APPROVED** [review]

File: src/saveload/gamelog_sl.cpp (hunk 4 above)
Comment (TrueBrain) [reply src/saveload/gamelog_sl.cpp:92]: Exactly the last part. The assert was there as the length of the list was important. That is no longer the case. And especially as you can now view a savegame, I think we have to assume the developer validates it :D

**Pull request merged**